src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.Infrastructure
+{
+    using ReaderFunc = Func<IAsyncEnumerable<object>, Task<ICollection>>;
+
+    /// <summary>
+    /// Type that reads an <see cref="IAsyncEnumerable{T}"/> instance into a
+    /// generic collection instance.
+    /// </summary>
+    /// <remarks>
+    /// This type is used to create a strongly typed synchronous <see cref="ICollection{T}"/> instance from
+    /// an <see cref="IAsyncEnumerable{T}"/>. An accurate <see cref="ICollection{T}"/> is required for XML formatters to 
+    /// correctly serialize.
+    /// </remarks>
+    public sealed class AsyncEnumerableReader
```

Comment:
Does this really need to be public? I don't have a strong objection, but it seems like infrastructure, rather than something we expect users to need.

Reply:
I could defer reading this from DI or have duplicated caching if your app uses a mix of `ObjectResult` and `JsonResult` returning `IAsyncEnumerable`. I guess the latter is rare and easily fixable if it turns out to be a problem in the future.

Reply:
I don't understand your answer. What's the reason why this type is public?

Reply:
`ObjectResultExecutor` is public. If we have to share the same singleton instance between the JsonResultExecutor(s) and ObjectResultExecutor, we either have to make the type public for ctor DI or read it in the body of `ExecuteAsync`.